<a href="https://colab.research.google.com/github/kaholict/Karpov-Course/blob/main/K_C_lesson_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

Ссылка на файл https://docs.google.com/spreadsheets/d/14eF-oNIwvEobXyUQf81hpO7urOoTS3I76xHlA0q1crg/edit?usp=sharing

In [ ]:
# !gdown --id 14eF-oNIwvEobXyUQf81hpO7urOoTS3I76xHlA0q1crg

pickup_dt – период с точностью до часа  
pickup_month – месяц  
borough – район Нью-Йорка, из которого был сделан  заказ (5 районов + аэропорт)  
pickups – число поездок за период (час)  
hday – является ли день праздничным/выходным; Y - да,  N - нет  
spd – скорость ветра в милях в час  
vsb – видимость  
temp – температура в градусах Фаренгейта  
dewp – точка росы по Фаренгейту  
slp – давление  
pcp_01 – количество осадков за час  
pcp_06 – количество осадков за 6 часов  
pcp_24 – количество осадков за 24 часа  
sd – глубина снега в дюймах  

In [ ]:
df = pd.read_csv('/content/2_taxi_nyc.csv')

In [ ]:
df.head()

pickup_dt pickup_month    borough  pickups hday  spd   vsb  temp  \
0  2015-01-01 01:00:00          Jan      Bronx      152    Y  5.0  10.0  30.0   
1  2015-01-01 01:00:00          Jan   Brooklyn     1519    Y  5.0  10.0  30.0   
2  2015-01-01 01:00:00          Jan        EWR        0    Y  5.0  10.0  30.0   
3  2015-01-01 01:00:00          Jan  Manhattan     5258    Y  5.0  10.0  30.0   
4  2015-01-01 01:00:00          Jan     Queens      405    Y  5.0  10.0  30.0   

   dewp     slp  pcp 01  pcp 06  pcp 24   sd  
0   7.0  1023.5     0.0     0.0     0.0  0.0  
1   7.0  1023.5     0.0     0.0     0.0  0.0  
2   7.0  1023.5     0.0     0.0     0.0  0.0  
3   7.0  1023.5     0.0     0.0     0.0  0.0  
4   7.0  1023.5     0.0     0.0     0.0  0.0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29101 entries, 0 to 29100
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   pickup_dt     29101 non-null  object 
 1   pickup_month  29101 non-null  object 
 2   borough       26058 non-null  object 
 3   pickups       29101 non-null  int64  
 4   hday          29101 non-null  object 
 5   spd           29101 non-null  float64
 6   vsb           29101 non-null  float64
 7   temp          29101 non-null  float64
 8   dewp          29101 non-null  float64
 9   slp           29101 non-null  float64
 10  pcp 01        29101 non-null  float64
 11  pcp 06        29101 non-null  float64
 12  pcp 24        29101 non-null  float64
 13  sd            29101 non-null  float64
dtypes: float64(9), int64(1), object(4)
memory usage: 3.1+ MB


Преобладающим типом данных в датасете является float64

In [ ]:
df.columns = [name.replace(' ', '_') for name in df.columns]

In [ ]:
df.columns

Index(['pickup_dt', 'pickup_month', 'borough', 'pickups', 'hday', 'spd', 'vsb',
       'temp', 'dewp', 'slp', 'pcp_01', 'pcp_06', 'pcp_24', 'sd'],
      dtype='object')

In [ ]:
df.head()

pickup_dt pickup_month    borough  pickups hday  spd   vsb  temp  \
0  2015-01-01 01:00:00          Jan      Bronx      152    Y  5.0  10.0  30.0   
1  2015-01-01 01:00:00          Jan   Brooklyn     1519    Y  5.0  10.0  30.0   
2  2015-01-01 01:00:00          Jan        EWR        0    Y  5.0  10.0  30.0   
3  2015-01-01 01:00:00          Jan  Manhattan     5258    Y  5.0  10.0  30.0   
4  2015-01-01 01:00:00          Jan     Queens      405    Y  5.0  10.0  30.0   

   dewp     slp  pcp_01  pcp_06  pcp_24   sd  
0   7.0  1023.5     0.0     0.0     0.0  0.0  
1   7.0  1023.5     0.0     0.0     0.0  0.0  
2   7.0  1023.5     0.0     0.0     0.0  0.0  
3   7.0  1023.5     0.0     0.0     0.0  0.0  
4   7.0  1023.5     0.0     0.0     0.0  0.0

Сколько записей (строк) в датафрейме относятся к району Манхэттен (Manhattan)?

In [ ]:
df.borough.value_counts()

Bronx            4343
Brooklyn         4343
EWR              4343
Manhattan        4343
Queens           4343
Staten Island    4343
Name: borough, dtype: int64

Сколько раз в данных встречается район Бруклин (Brooklyn)?  
Бруклин встречается 4343 раза.

In [ ]:
df.pickups.sum()

14265773

Общее количество поездок равняется 14265773

In [ ]:
ccc = df.groupby('borough', as_index = False).agg({'pickups': 'sum'})
ccc

borough   pickups
0          Bronx    220047
1       Brooklyn   2321035
2            EWR       105
3      Manhattan  10367841
4         Queens   1343528
5  Staten Island      6957

In [ ]:
ccc.pickups.sum()

14259513

Наибольшее количество поездок было совершено из района Manhattan

Найдём район с минимальным числом поездок с помoщью функции idxmin():

In [ ]:
min_pickups = df.groupby('borough', as_index = False).agg({'pickups': 'sum'}).borough[df.groupby('borough', as_index = False).agg({'pickups': 'sum'}).pickups.idxmin()]

In [ ]:
min_pickups

'EWR'

*Сгруппируем данные по двум признакам: району города и является ли день выходным (колонки borough и hday).  
Сравним среднее число поездок, и назовите районы, из которых по праздникам в среднем поступает больше заказов, чем в обычные дни.*

In [ ]:
df.head()

pickup_dt pickup_month    borough  pickups hday  spd   vsb  temp  \
0  2015-01-01 01:00:00          Jan      Bronx      152    Y  5.0  10.0  30.0   
1  2015-01-01 01:00:00          Jan   Brooklyn     1519    Y  5.0  10.0  30.0   
2  2015-01-01 01:00:00          Jan        EWR        0    Y  5.0  10.0  30.0   
3  2015-01-01 01:00:00          Jan  Manhattan     5258    Y  5.0  10.0  30.0   
4  2015-01-01 01:00:00          Jan     Queens      405    Y  5.0  10.0  30.0   

   dewp     slp  pcp_01  pcp_06  pcp_24   sd  
0   7.0  1023.5     0.0     0.0     0.0  0.0  
1   7.0  1023.5     0.0     0.0     0.0  0.0  
2   7.0  1023.5     0.0     0.0     0.0  0.0  
3   7.0  1023.5     0.0     0.0     0.0  0.0  
4   7.0  1023.5     0.0     0.0     0.0  0.0

In [ ]:
# compare_in_hd = df.query('hday == "Y"').groupby(['borough'], as_index=False).agg({'pickups': sum})

In [ ]:
# compare_in_hd["mean_in_day"] = [compare_in_hd.pickups[i] / 2 for i in range(len(compare_in_hd))]

In [ ]:
# compare_in_hd

In [ ]:
# compare_not_in_hd = df.query('hday == "N"').groupby(['borough'], as_index=False).agg({'pickups': sum})

In [ ]:
# compare_not_in_hd["mean_in_day"] = [compare_not_in_hd.pickups[i] / 5 for i in range(len(compare_not_in_hd))]

In [ ]:
# compare_not_in_hd

In [ ]:
# df['hday'] = df['hday'].fillna('no_info')

In [ ]:
# t = df.groupby(['borough', 'hday'], as_index=False).agg({'pickups': ['sum']})

In [ ]:
# t.columns = ['_'.join(col).strip() for col in t.columns.values]

In [ ]:
# t["mean_in_day"] = [t.pickups_sum[i] / 2 if t.hday_[i] == 'Y' else t.pickups_sum[i] / 5 for i in range(len(t))]
# t

In [ ]:
# if t.pickups_sum.sum() == df.pickups.sum():
#   print('СХОДИТСЯ!!!!')
# else:
#   print('НЕ СХОДИТСЯ(((')
#   print(t.pickups_sum.sum(), df.pickups.sum())

In [ ]:
t = df.groupby(['borough', 'hday'], as_index=False).agg({'pickups': 'mean'})
t.columns

Index(['borough', 'hday', 'pickups'], dtype='object')

In [ ]:
t

borough hday      pickups
0           Bronx    N    50.771073
1           Bronx    Y    48.065868
2        Brooklyn    N   534.727969
3        Brooklyn    Y   527.011976
4             EWR    N     0.023467
5             EWR    Y     0.041916
6       Manhattan    N  2401.302921
7       Manhattan    Y  2035.928144
8          Queens    N   308.899904
9          Queens    Y   320.730539
10  Staten Island    N     1.606082
11  Staten Island    Y     1.497006

In [ ]:
more_in_holly = [t.borough[i] for i in range(1, len(t), 2) if t.pickups[i] > t.pickups[i-1]]
more_in_holly

['EWR', 'Queens']

Из районов EWR и Queens поступает больше заказов в выходные дни, чем в будние.

*Для каждого района посчитайте число поездок по месяцам.  
Отсортируйте полученные значения по убыванию и сохраните результирующий датафрейм в pickups_by_mon_bor.  
Обратите внимание, что итоговый датасет должен состоять из 3-х колонок – pickup_month, borough, pickups.*

In [ ]:
pickup_month = df.groupby(['borough', 'pickup_month']).agg({'pickups': 'sum'})
pickup_month

pickups
borough       pickup_month         
Bronx         Apr             34617
              Feb             28694
              Jan             22461
              Jun             49006
              Mar             32232
              May             53037
Brooklyn      Apr            378095
              Feb            328650
              Jan            309011
              Jun            482466
              Mar            346726
              May            476087
EWR           Apr                10
              Feb                14
              Jan                11
              Jun                29
              Mar                14
              May                27
Manhattan     Apr           1648278
              Feb           1718571
              Jan           1455543
              Jun           1995388
              Mar           1661261
              May           1888800
Queens        Apr            216857
              Feb            185695
              Jan            159211
              Jun            286311
              Mar            219561
              May            275893
Staten Island Apr              1068
              Feb               903
              Jan               821
              Jun              1673
              Mar               975
              May              1517

In [ ]:
def temp_to_celcius(col):
  Celsius_arr = []
  for Fahrenheit in col:
    Celsius = ((Fahrenheit-32) *5.0)/(9.0)
    Celsius_arr.append(Celsius)
  return Celsius_arr

In [ ]:
df['temp_C'] = temp_to_celcius(df['temp'])

In [ ]:
df['temp_C'][:5]

0   -1.111111
1   -1.111111
2   -1.111111
3   -1.111111
4   -1.111111
Name: temp_C, dtype: float64